include/deal.II/matrix_free/cuda_fe_evaluation.h

```diff
@@ -40,561 +25,7 @@ DEAL_II_NAMESPACE_OPEN
-  }
-
-
-
-  template <int dim,
-            int fe_degree,
-            int n_q_points_1d,
-            int n_components_,
-            typename Number>
-  DEAL_II_HOST_DEVICE void
-  FEEvaluation<dim, fe_degree, n_q_points_1d, n_components_, Number>::
-    submit_dof_value(const value_type &val_in, int q_point)
-  {
-    shared_data->values(q_point) = val_in;
-  }
-
-
-
-  template <int dim,
-            int fe_degree,
-            int n_q_points_1d,
-            int n_components_,
-            typename Number>
-  DEAL_II_HOST_DEVICE typename FEEvaluation<dim,
-                                            fe_degree,
-                                            n_q_points_1d,
-                                            n_components_,
-                                            Number>::gradient_type
-  FEEvaluation<dim, fe_degree, n_q_points_1d, n_components_, Number>::
-    get_gradient(int q_point) const
-  {
-    static_assert(n_components_ == 1, "This function only supports FE with one \
-                  components");
-
-    gradient_type grad;
-    for (unsigned int d_1 = 0; d_1 < dim; ++d_1)
-      {
-        Number tmp = 0.;
-        for (unsigned int d_2 = 0; d_2 < dim; ++d_2)
-          tmp += data->inv_jacobian(cell_id, q_point, d_2, d_1) *
-                 shared_data->gradients(q_point, d_2);
-        grad[d_1] = tmp;
-      }
-
-    return grad;
-  }
-
-
-
-  template <int dim,
-            int fe_degree,
-            int n_q_points_1d,
-            int n_components_,
-            typename Number>
-  DEAL_II_HOST_DEVICE void
-  FEEvaluation<dim, fe_degree, n_q_points_1d, n_components_, Number>::
-    submit_gradient(const gradient_type &grad_in, int q_point)
-  {
-    for (unsigned int d_1 = 0; d_1 < dim; ++d_1)
-      {
-        Number tmp = 0.;
-        for (unsigned int d_2 = 0; d_2 < dim; ++d_2)
-          tmp += data->inv_jacobian(cell_id, q_point, d_1, d_2) * grad_in[d_2];
-        shared_data->gradients(q_point, d_1) =
-          tmp * data->JxW(cell_id, q_point);
-      }
-  }
-
-
-
-  template <int dim,
-            int fe_degree,
-            int n_q_points_1d,
-            int n_components_,
-            typename Number>
-  template <typename Functor>
-  DEAL_II_HOST_DEVICE void
-  FEEvaluation<dim, fe_degree, n_q_points_1d, n_components_, Number>::
-    apply_for_each_quad_point(const Functor &func)
-  {
-    Kokkos::parallel_for(Kokkos::TeamThreadRange(shared_data->team_member,
-                                                 n_q_points),
-                         [&](const int &i) { func(this, i); });
-    shared_data->team_member.team_barrier();
-  }
-
-
-
-#ifndef DOXYGEN
-  template <int dim,
-            int fe_degree,
-            int n_q_points_1d,
-            int n_components_,
-            typename Number>
-  constexpr unsigned int
-    FEEvaluation<dim, fe_degree, n_q_points_1d, n_components_, Number>::
-      n_q_points;
-#endif
+  using namespace Portable;
```

Comment:
That's clever. Is there also a way to effectively deprecate the whole `CUDAWrappers` namespace?

Reply:
The problem is that `nvcc` doesn't like when I deprecate the namespace. With the others compilers I can do
`namespace DEAL_II_EARLY_DEPRECATION CUDAWrappers` and it works like expected. We could only emit the deprecation warning when people use a compiler other than `nvcc` but I would expect that almost everybody that uses this class also uses `nvcc`.

Reply:
What does "`nvcc` doesn't like when I deprecate the namespace` exactly mean?
Is the attribute just not recognized and the compiler warns about that (like in https://godbolt.org/z/Mf3444sP9 which only recognizes the attribute from `nvcc 11.1.0` on) or is there a different issue?

Reply:
Yes, the compiler complains that attribute is not recognized.

Reply:
We should just do a version check then since it seems to work from `nvcc 11.1.0` on.

Reply:
I don't think that the issue is due to `nvcc` version.  The CI fails with `nvcc` 11.8. I think that the problem is the underlying gcc compiler.  The CI uses gcc 9 which had a bug when using `[[deprecated]]` with a namespace (https://gcc.gnu.org/bugzilla/show_bug.cgi?id=79817)